cli/upgrade.rs

```diff
@@ -0,0 +1,189 @@
+    "https://github.com/denoland/deno/releases/download/v".to_string();
+  static ref REGEX_STRING: String = r#"v([^\?]+)?""#.to_string();
+  static ref DENO_EXEC_TEMP_NAME: String = "deno_temp".to_string();
+}
+
+#[cfg(windows)]
+const EXEC_FILE_NAME: &str = "deno_win_x64.zip";
+#[cfg(macos)]
+const EXEC_FILE_NAME: &str = "deno_osx_x64.gz";
+#[cfg(unix)]
+const EXEC_FILE_NAME: &str = "deno_linux_x64.gz";
+
+struct ErrorMsg(String);
+
+impl ErrorMsg {
+  fn to_err_box(&self) -> ErrBox {
+    ErrBox::from(std::io::Error::new(
+      std::io::ErrorKind::Other,
+      self.0.clone(),
+    ))
+  }
+}
+
+/// Asynchronously updates deno executable to greatest version
+/// if greatest version is available.
+pub async fn exec_upgrade() -> Result<(), ErrBox> {
+  let client = Client::builder().redirect(Policy::none()).build()?;
+  println!("Checking for latest version");
+  let body = client
+    .get(Url::parse(&LATEST_VERSION_URL)?)
+    .send()
+    .await?
+    .text()
+    .await?;
+  let checked_version = find_version(&body)?;
+  if is_latest_version_greater(&version::DENO.to_string(), &checked_version) {
+    println!(
+      "New version has been found\nDeno is upgrading to version {}",
+      &checked_version
+    );
+    let archive =
+      download_package(&compose_url_to_exec(&checked_version)?, client).await?;
+    let path = current_exe()?;
+    unpack(archive, &path)?;
+    replace_exec(&path)?;
+    println!("Upgrade done successfully")
+  } else {
+    println!("Local deno version {} is the greatest one", &version::DENO);
+  }
+  Ok(())
+}
+
+fn download_package(
+  url: &Url,
+  client: Client,
+) -> Pin<Box<dyn Future<Output = Result<Vec<u8>, ErrBox>>>> {
+  let url = url.clone();
+  let fut = async move {
+    match fetch_once(client.clone(), &url, None).await? {
+      FetchOnceResult::Code(source, _) => Ok(source),
+      FetchOnceResult::NotModified => Err(
+        ErrorMsg(format!("Cannot fetch data from url: {}", &url)).to_err_box(),
+      ),
+      FetchOnceResult::Redirect(_url, _) => {
+        download_package(&_url, client).await
+      }
+    }
+  };
+  fut.boxed_local()
+}
+
+fn compose_url_to_exec(version: &str) -> Result<Url, ErrBox> {
+  let mut url_str = EXEC_DOWNLOAD_URL.clone();
+  url_str.push_str(&format!("{}/", version));
+  url_str.push_str(&EXEC_FILE_NAME);
+  Ok(Url::parse(&url_str[..])?)
+}
+
+fn find_version(text: &str) -> Result<String, ErrBox> {
+  let re = Regex::new(&REGEX_STRING)?;
+  if let Some(_mat) = re.find(text) {
+    let mat = _mat.as_str();
+    return Ok(mat[1..mat.len() - 1].to_string());
+  }
+  Err(ErrorMsg("Cannot read latest tag version".to_string()).to_err_box())
+}
+
+fn is_latest_version_greater(old_v: &str, new_v: &str) -> bool {
+  let mut power = 4;
+  let (mut old_v_num, mut new_v_num) = (0, 0);
+  old_v
+    .split('.')
+    .zip(new_v.split('.'))
+    .for_each(|(old, new)| {
+      old_v_num += old.parse::<i32>().unwrap() * (10_f32.powi(power) as i32);
```

Comment:
We probably don't need the hack as we follow the semver

Reply:
semver is a next crate to use, shall I do so? thx for suggestion.

Reply:
I meant it seems you are doing the hacky number manipulation here. Following semvar means we can just compare number at each position directly